### ISSUE TITLE
Pipeline restarted mid-step cannot resume

### ISSUE THREAD

**Issue by chrishiner** [jenkinsci/templating-engine-plugin#191]
### Environment Information

- [x] Jenkins version 2.277.2 (running in K8s)
- [x] Plugin version 2.1.0
- [x] OS Debian 10
All other plugins are running the most recently available versions as of today.

### Description

If I restart jenkins while the sleep command in my test step is running, it tries to resume, but throws an exception.

### Steps to Reproduce

1. Create a step with a sleep:
```
void call(){
  println "running before sleep"
  sleep 120
  println "running after sleep"
}
```
2. run the step in a template
3. restart jenkins during the sleep

<details>
<summary>Log</summary>

```
[JTE] Obtained Pipeline Configuration File
[JTE] -- scm: git https://XXXX
[JTE] -- file path: pipeline_config.groovy
[JTE] Pipeline Configuration Modifications
[JTE] Configurations Added:
[JTE] - jte.allow_scm_jenkinsfile set to false
[JTE] - jte.pipeline_template set to testPipeline
[JTE] - libraries set to [:]
[JTE] - application_environments.test.long_name set to Test
[JTE] - template_methods set to [:]
[JTE] Configurations Deleted: None
[JTE] Configurations Changed: None
[JTE] Configurations Duplicated: None
[JTE] Configurations Ignored: None
[JTE] Subsequent May Merge:
[JTE] - jte
[JTE] - libraries
[JTE] - application_environments
[JTE] - application_environments.test
[JTE] Subsequent May Override:
[JTE] - jte.pipeline_template
[JTE] Obtained Template Configuration File
[JTE] -- scm: git https://XXXX
[JTE] -- file path: pipeline_config.groovy
[JTE] Pipeline Configuration Modifications
[JTE] Configurations Added:
[JTE] - libraries.jte_error_tester set to [:]
[JTE] Configurations Deleted:
[JTE] - libraries
[JTE] Configurations Changed: None
[JTE] Configurations Duplicated: None
[JTE] Configurations Ignored:
[JTE] - application_environments
[JTE] Subsequent May Merge: None
[JTE] Subsequent May Override: None
[JTE] Loading Library jte_error_tester
[JTE] -- scm: git https://XXXX
[JTE] created Application Environments:
[JTE] test
[JTE] created Library Steps:
[JTE] withResumeError from the jte_error_tester Library
[JTE] Obtained Repository Jenkinsfile
[JTE] -- scm: git https://XXXX
[JTE] -- file path: Jenkinsfile
[JTE] Repository provided Jenkinsfile that will not be used, per organizational policy.
[JTE] Obtained Pipeline Template
[JTE] -- scm: git https://XXXX
[JTE] -- file path: pipeline_templates/testPipeline
Running in Durability level: MAX_SURVIVABILITY
[Pipeline] Start of Pipeline
[Pipeline] stage
[Pipeline] { (Test)
[JTE][Step - jte_error_tester/withResumeError.call()]
[Pipeline] echo
withResumeError before
[Pipeline] echo
New sleep start
[Pipeline] sleep
Sleeping for 2 min 0 sec
Resuming build at Mon Mar 29 10:40:27 EDT 2021 after Jenkins restart
[Pipeline] End of Pipeline
an exception which occurred:
	in object of type java.util.HashMap
	in field com.cloudbees.groovy.cps.impl.FunctionCallEnv.locals
	in object com.cloudbees.groovy.cps.impl.FunctionCallEnv@44a6de4c
	in object of type com.cloudbees.groovy.cps.impl.FunctionCallEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.BlockScopeEnv@23e71d8b
	in object of type com.cloudbees.groovy.cps.impl.BlockScopeEnv
	in field com.cloudbees.groovy.cps.impl.ProxyEnv.parent
	in object com.cloudbees.groovy.cps.impl.BlockScopeEnv@191a4d9d
	in object of type com.cloudbees.groovy.cps.impl.BlockScopeEnv
	in field com.cloudbees.groovy.cps.impl.CallEnv.caller
	in object com.cloudbees.groovy.cps.impl.FunctionCallEnv@15b74fc7
	in object of type com.cloudbees.groovy.cps.impl.FunctionCallEnv
	in field com.cloudbees.groovy.cps.Continuable.e
	in object org.jenkinsci.plugins.workflow.cps.SandboxContinuable@6791a783
	in object of type org.jenkinsci.plugins.workflow.cps.SandboxContinuable
	in field org.jenkinsci.plugins.workflow.cps.CpsThread.program
	in object org.jenkinsci.plugins.workflow.cps.CpsThread@7c81e5bf
	in object of type org.jenkinsci.plugins.workflow.cps.CpsThread
	in object of type java.util.concurrent.ConcurrentSkipListMap
	in field org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.threads
	in object org.jenkinsci.plugins.workflow.cps.CpsThreadGroup@373c3917
	in object of type org.jenkinsci.plugins.workflow.cps.CpsThreadGroup
Caused: java.io.InvalidClassException: WorkflowScript; Class does not extend stream superclass
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1046)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1355)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.BlockUnmarshaller.readObject(BlockUnmarshaller.java:149)
	at org.jboss.marshalling.river.BlockUnmarshaller.readObject(BlockUnmarshaller.java:135)
	at org.jboss.marshalling.MarshallerObjectInputStream.readObjectOverride(MarshallerObjectInputStream.java:53)
	at org.jboss.marshalling.river.RiverObjectInputStream.readObjectOverride(RiverObjectInputStream.java:307)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:494)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:461)
	at java.util.HashMap.readObject(HashMap.java:1412)
	at sun.reflect.GeneratedMethodAccessor38.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.marshalling.reflect.JDKSpecific$SerMethods.callReadObject(JDKSpecific.java:179)
	at org.jboss.marshalling.reflect.SerializableClass.callReadObject(SerializableClass.java:212)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1746)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1715)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1715)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1715)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1715)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.BlockUnmarshaller.readObject(BlockUnmarshaller.java:149)
	at org.jboss.marshalling.river.BlockUnmarshaller.readObject(BlockUnmarshaller.java:135)
	at org.jboss.marshalling.MarshallerObjectInputStream.readObjectOverride(MarshallerObjectInputStream.java:53)
	at org.jboss.marshalling.river.RiverObjectInputStream.readObjectOverride(RiverObjectInputStream.java:307)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:494)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:461)
	at java.util.concurrent.ConcurrentSkipListMap.readObject(ConcurrentSkipListMap.java:1481)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.marshalling.reflect.JDKSpecific$SerMethods.callReadObject(JDKSpecific.java:179)
	at org.jboss.marshalling.reflect.SerializableClass.callReadObject(SerializableClass.java:212)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1746)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:220)
	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1853)
	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1767)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1395)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:272)
	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:205)
	at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41)
	at org.jenkinsci.plugins.workflow.support.pickles.serialization.RiverReader$SandboxedUnmarshaller.lambda$readObject$0(RiverReader.java:250)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:237)
	at org.jenkinsci.plugins.workflow.support.pickles.serialization.RiverReader$SandboxedUnmarshaller.sandbox(RiverReader.java:237)
	at org.jenkinsci.plugins.workflow.support.pickles.serialization.RiverReader$SandboxedUnmarshaller.readObject(RiverReader.java:250)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$2.onSuccess(CpsFlowExecution.java:793)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$2.onSuccess(CpsFlowExecution.java:786)
	at org.jenkinsci.plugins.workflow.support.concurrent.Futures$1.run(Futures.java:153)
	at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:253)
	at com.google.common.util.concurrent.ExecutionList$RunnableExecutorPair.execute(ExecutionList.java:149)
	at com.google.common.util.concurrent.ExecutionList.add(ExecutionList.java:105)
	at com.google.common.util.concurrent.AbstractFuture.addListener(AbstractFuture.java:155)
	at org.jenkinsci.plugins.workflow.support.concurrent.Futures.addCallback(Futures.java:163)
	at org.jenkinsci.plugins.workflow.support.concurrent.Futures.addCallback(Futures.java:93)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.loadProgramAsync(CpsFlowExecution.java:783)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:750)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:691)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:550)
	at hudson.model.RunMap.retrieve(RunMap.java:225)
	at hudson.model.RunMap.retrieve(RunMap.java:57)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:501)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:483)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:381)
	at hudson.model.RunMap.getById(RunMap.java:205)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.run(WorkflowRun.java:940)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.get(WorkflowRun.java:951)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$1.computeNext(FlowExecutionList.java:65)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$1.computeNext(FlowExecutionList.java:57)
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
	at org.jenkinsci.plugins.workflow.flow.FlowExecutionList$ItemListenerImpl.onLoaded(FlowExecutionList.java:178)
	at jenkins.model.Jenkins.<init>(Jenkins.java:1015)
	at hudson.model.Hudson.<init>(Hudson.java:85)
	at hudson.model.Hudson.<init>(Hudson.java:81)
	at hudson.WebAppMain$3.run(WebAppMain.java:295)
Caused: java.io.IOException: Failed to load build state
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$3.onSuccess(CpsFlowExecution.java:865)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$3.onSuccess(CpsFlowExecution.java:863)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$4$1.run(CpsFlowExecution.java:917)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$1.run(CpsVmExecutorService.java:38)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:139)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

</details>

**Comment by steven-terrana** [comment]
Thanks for letting us know, we'll look into it and see if we can reproduce/fix.

**Comment by chrishiner** [comment]
If the pipeline template file contains just "sleep 120" then it resumes as expected when jenkins is restarted.
When I change it to call it via the step "withResumeError()" then it fails.

**Comment by chrishiner** [comment]
That Class does not extend stream superclass error is thrown here:
https://github.com/jboss-remoting/jboss-marshalling/blob/2.0.6.Final/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java#L1046
At the point when it fails it's reading from the program.dat file and:
superType=class org.boozallen.plugins.jte.init.primitives.injectors.StepWrapperScript
superDescriptor=class org.boozallen.plugins.jte.init.primitives.injectors.StepWrapperScript

I'm not sure why it's able to write the file, but then not read it.
I dug around a bit more, but I don't have enough understanding of the Cps magic to debug much further.

**Comment by cokieffebah** [comment]
Thanks we are seeing a similar error in some other work we are doing

**Comment by Opa-** [comment]
Hi,

Just to report that we are seeing the same issue with our environment being:

- [x] Jenkins 2.298
- [x] JTE 2.2.2 
- [x] OS Debian 10

**Comment by kjayaraman-fiserv** [comment]
Hi, we have a simiar issue in our environment aswell. Jenkins 2.263.4, JTE 2.2.2, RHEL 6

**Comment by Opa-** [comment]
Still experiencing the issue with following environment:

- [x] Jenkins 2.231
- [x] JTE 2.3
- [x] OS Debian 11

**Comment by steven-terrana** [comment]
have been able to reproduce this via a unit test:
https://github.com/steven-terrana/templating-engine-plugin/blob/fix-191/src/test/groovy/org/boozallen/plugins/jte/init/ResumabilitySpec.groovy#L134-L155

haven't gotten much further than @chrishiner though in troubleshooting. 

this is being caused by a failure of Jenkins to unmarshal / rehydrate a JTE step. 

The current rabbit hole i'm investigating is figuring out how a Jenkinsfile, which gets compiled with [CpsScript](https://github.com/jenkinsci/workflow-cps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScript.java) as its Script Base Class, is able to rehydrate successfully. 

JTE step's are compiled in a similar manner, with [StepWrapperScript](https://github.com/jenkinsci/templating-engine-plugin/blob/main/src/main/java/org/boozallen/plugins/jte/init/primitives/injectors/StepWrapperScript.java) being analogous to CpsScript for steps. 

_somehow_ the parsed class from a Jenkinsfile ("WorkflowScript") is assignable from CpsScript but not from StepWrapperScript.

**Comment by steven-terrana** [comment]
fixed by #273! 🥳
